### PR TITLE
ENH: Add GeneratedBy and SourceDatasets to dataset_description schema

### PR DESCRIPTION
**Pull request by effigies**
Need to make `GeneratedBy` mandatory if `DatasetType` is `"derivative"`. Not sure where yet. And should warn if `SourceDatasets` is missing.

### CODE REVIEW

**Comment by sappelhoff** [comment]
> Need to make GeneratedBy mandatory if DatasetType is "derivative". Not sure where yet

you can copy the way that Ephys *requires* `*Description` if `*System` is "Other"

https://github.com/bids-standard/bids-validator/blob/c1ecfba041adcc842ab11dee4c78c3e7155e3fae/bids-validator/validators/json/schemas/coordsystem_eeg.json#L20-L57

**Comment by effigies** [comment]
Okay, at least got to the point where the tests I added are failing. Hopefully a reviewer will see why they're failing.

File: bids-validator/validators/json/schemas/dataset_description.json [hunk 1]
```diff
@@ -70,5 +103,19 @@
     }
   },
   "required": ["Name", "BIDSVersion"],
-  "type": "object"
+  "allOf": [
+    { "$ref": "#/dependency-definitions/if-DatasetTpe-is-derivative-then-GeneratedBy-is-required" }
```
Comment (sappelhoff) [inline bids-validator/validators/json/schemas/dataset_description.json]:
```suggestion
    { "$ref": "#/dependency-definitions/if-DatasetType-is-derivative-then-GeneratedBy-is-required" }
```

File: bids-validator/validators/json/schemas/dataset_description.json [hunk 2]
```diff
@@ -70,5 +103,19 @@
...
+  ],
+  "dependency-definitions": {
+    "if-DatasetTpe-is-derivative-then-GeneratedBy-is-required": {
```
Comment (sappelhoff) [inline bids-validator/validators/json/schemas/dataset_description.json]:
```suggestion
    "if-DatasetType-is-derivative-then-GeneratedBy-is-required": {
```

**Comment by sappelhoff** [comment]
> Okay, at least got to the point where the tests I added are failing. Hopefully a reviewer will see why they're failing.

I cannot say exactly but in a similar situation I have started to add "console.log" to the test code, to find where something is wrong. You might already have done so, please ignore my comment in that case.

I would add "console.log(issues)" one line before the current lines 524 and 537 in your tests.

File: bids-validator/tests/json.spec.js [hunk 3]
```diff
@@ -496,4 +496,45 @@ describe('JSON', function() {
...
+    var jsonObj = {
+      Name: 'Example Name',
+      BIDSVersion: '1.4.0',
+      Authors: ['example author'],
+      DatasetType: 'derivative',
+      GeneratedBy: [],
+    }
+    jsonDict[dataset_description_file.relativePath] = jsonObj
+    validate.JSON(dataset_description_file, jsonDict, function(issues) {
+      assert(issues.length === 1 && issues[0].code == 55)
```
Comment (sappelhoff) [inline bids-validator/tests/json.spec.js]: I have a suspicion that you have more than 1 issue here. (that is, length === 2 or more)

compare to the ephys "conditional requirement" test: https://github.com/bids-standard/bids-validator/blob/c1ecfba041adcc842ab11dee4c78c3e7155e3fae/bids-validator/tests/json.spec.js#L254-L273

**Review by sappelhoff — APPROVED** [review]
LGTM apart from linting fixes.

I haven't seen a straight forward way to *warn* using JSON schema, but Ross linked this at the last maintainers meeting: https://github.com/ajv-validator/ajv-errors

**Review by rwblair — APPROVED** [review]

**Comment by rwblair** [comment]
@sappelhoff  I'm looking into modifying the approach used by ajv-errors to issue warnings, haven't found another pure schema way. Would be very helpful for the ASL validator changes too, but I have no time line on getting it working.

**Comment by sappelhoff** [comment]
Cool, let me know if you have something that I should try/test :)

**Pull request merged**